picard/util/pipe.py

```diff
@@ -0,0 +1,279 @@
+        if args is None:
+            args = tuple()
+        elif isinstance(args, str):
+            args = (args,)
+        else:
+            try:
+                args = tuple(args)
+            except TypeError as exc:
+                raise PipeErrorInvalidArgs(exc)
+
+        if not args:
+            args = (self.MESSAGE_TO_IGNORE,)
+
+        self.__app_name: str = app_name
+        self.__app_version: str = app_version
+        self.__is_mac: bool = IS_MACOS
+        self.__is_win: bool = IS_WIN
+
+        # named pipe values needed by windows API
+        if self.__is_win:
+            self.__app_version = self.__app_version.replace(".", "-")
+
+            # win32pipe.CreateNamedPipe
+            # more about the arguments: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea
+            self.__MAX_INSTANCES: int = 1
+            self.__BUFFER_SIZE: int = 65536
+            # timeout doesn't really matter, concurrent.futures ensures that connections are closed in declared time
+            # the value is in milliseconds
+            self.__DEFAULT_TIMEOUT: int = 300
+
+            # win32file.CreateFile
+            # more about the arguments: http://timgolden.me.uk/pywin32-docs/win32file__CreateFile_meth.html
+            self.__SHARE_MODE: int = 0
+            self.__FLAGS_AND_ATTRIBUTES: int = 0
+
+            # pywintypes.error error codes
+            # more about the error codes: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/18d8fbe8-a967-4f1c-ae50-99ca8e491d2d
+            self.__FILE_NOT_FOUND_ERROR_CODE: int = 2
+            self.__BROKEN_PIPE_ERROR_CODE: int = 109
+
+        self.path: str = self.__generate_filename()
+
+        self.is_pipe_owner: bool = False
+        self.permission_error_happened: bool = False
+
+        if self.__is_win:
+            for arg in args:
+                if not self.send_to_pipe(arg):
+                    self.is_pipe_owner = True
+                    break
+        else:
+            try:
+                self.__create_unix_pipe()
+            except FileExistsError:
+                for arg in args:
+                    if not self.send_to_pipe(arg):
+                        self.__create_unix_pipe()
+                        break
+
+    def __generate_filename(self) -> str:
+        if self.__is_win:
+            self.__pipe_parent_dir = "\\\\.\\pipe\\"
+        elif self.__is_mac:
+            self.__pipe_parent_dir = os.path.expanduser("~/Library/Application Support/MusicBrainz/Picard/pipes/")
+        else:
+            self.__pipe_parent_dir = f"{os.getenv('XDG_RUNTIME_DIR')}/"
+            # just in case the $XDG_RUNTIME_DIR is not declared, fallback dir
+            if not self.__pipe_parent_dir:
+                self.__pipe_parent_dir = os.path.expanduser("~/.config/MusicBrainz/Picard/pipes/")
+
+        return f"{self.__pipe_parent_dir}{self.__app_name}_v{self.__app_version}_pipe_file"
```

Comment:
You can't build paths this way. Either use [os.path.join()](https://docs.python.org/3/library/os.path.html#os.path.join) or [PurePath.joinpath()](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.joinpath)

Reply:
> You can't build paths this way. Either use [os.path.join()](https://docs.python.org/3/library/os.path.html#os.path.join) or [PurePath.joinpath()](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.joinpath)

What line is wrong? 151 or 153?